hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/VolumeHealthMetrics.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.container.common.volume;
+
+import java.util.List;
+import org.apache.hadoop.metrics2.MetricsCollector;
+import org.apache.hadoop.metrics2.MetricsInfo;
+import org.apache.hadoop.metrics2.MetricsRecordBuilder;
+import org.apache.hadoop.metrics2.MetricsSource;
+import org.apache.hadoop.metrics2.MetricsSystem;
+import org.apache.hadoop.metrics2.annotation.Metrics;
+import org.apache.hadoop.metrics2.lib.DefaultMetricsSystem;
+import org.apache.hadoop.metrics2.lib.Interns;
+import org.apache.hadoop.metrics2.lib.MetricsRegistry;
+import org.apache.hadoop.ozone.OzoneConsts;
+
+/**
+ * This class is used to track Volume Health metrics for all volumes on a datanode.
+ */
+@Metrics(about = "Ozone Volume Health Metrics",
+    context = OzoneConsts.OZONE)
+public class VolumeHealthMetrics implements MetricsSource {
+
+  private static final String SOURCE_BASENAME =
+      VolumeHealthMetrics.class.getSimpleName();
+
+  private static final MetricsInfo TOTAL_VOLUMES =
+      Interns.info("TotalVolumes", "Total number of volumes");
+  private static final MetricsInfo HEALTHY_VOLUMES =
+      Interns.info("NumHealthyVolumes", "Number of healthy volumes");
+  private static final MetricsInfo FAILED_VOLUMES =
+      Interns.info("NumFailedVolumes", "Number of failed volumes");
+
+  private final MetricsRegistry registry;
+  private final String metricsSourceName;
+  private final VolumeSet volumeSet;
+  private final StorageVolume.VolumeType volumeType;
+
+  /**
+   * Constructor for VolumeHealthMetrics.
+   *
+   * @param volumeType Type of volumes (DATA_VOLUME, META_VOLUME, DB_VOLUME)
+   * @param volumeSet The volume set to track metrics for
+   */
+  public VolumeHealthMetrics(StorageVolume.VolumeType volumeType, VolumeSet volumeSet) {
+    this.volumeType = volumeType;
+    this.volumeSet = volumeSet;
```

Comment:
The composition here is backwards. `MutableVolumeSet` should contain the metrics object that tracks its stats. Then internally the locks can be used to get metrics from its maps as a consistent view in constant time. The only `OzoneContainer` modification would be to call a new `unregister` method in `MutableVolumeSet` to clean up its metrics on shutdown.